akka-docs/src/main/paradox/typed/actors.md

```diff
@@ -412,6 +412,28 @@ former simply speaks more languages than the latter. The opposite would be
 problematic, so passing an @scala[`ActorRef[PublishSessionMessage]`]@java[`ActorRef<PublishSessionMessage>`] where
 @scala[`ActorRef[RoomCommand]`]@java[`ActorRef<RoomCommand>`] is required will lead to a type error.
 
+@@@ div {.group-java}
+#### AbstractOnMessageBehavior API
+
+The `AbstractBehavior` API makes use of builders like `ReceiveBuilder`.  The benefit of these builders is that
+`instanceof` checks and casts are performed "behind the scenes", but this imposes some runtime overhead.  Pattern-matching
+features introduced in Java 17 and refined in subsequent versions allow the Java compiler to perform some of the
+work of the builders at compile-time, eliminating this runtime overhead without an increase in verbosity.  Users
+of earlier Java versions may also find eliminating this runtime overhead to be worth the verbosity of explicitly
+performing `instanceof` checks and casts.  Users of other JVM languages with pattern-matching support (such as Kotlin) may
+also prefer not using builders (note that the Scala DSL's `AbstractBehavior` does not make use of builders).
```

Comment:
If we have not benchmarked and actually seen a clear throughput improvement (and maybe not anyway), I think we should skip the talk about performance in the docs and just focus on the ergonomics.

Reply:
yes, maybe just say: directly patten matching maybe more performant if it's directly supported.

Reply:
Perhaps after this is merged it might be worth a benchmark and then mentioning the benchmark in the 2.7 release (maybe not the notes, but a blog after the release)?

@He-Pin Performance probably doesn't really have anything directly to do with the pattern matching.  Any performance win is attainable (modulo JIT and GC subtleties) with `AbstractOnMessageBehavior` in Java 8, since the pattern matching code AFAICT will basically compile into the same `instanceof` check and then cast (I wonder if anybody has attempted to decompile Java 17 pattern matching (perhaps manipulating the class file versions to say this was from Java 8) with a Java 8 emitting decompiler).

```java
if (msg instanceof GetSession) { return onGetSession((GetSession)msg); }
```

is a lot uglier than

```java
.onMessage(GetSession.class, this::onGetSession)
```

while

```java
case GetSession gs: return onGetSession(gs);
```

is at least in the same aesthetic ballpark, so if the first (in an `AbstractOnMessageBehavior`) is faster than the second and you're able to do the third, you're not giving up much/any aesthetics/understandability in order to get the performance win.

Pre-registering my hypothesis for the benchmarking run: `AbstractOnMessageBehavior` has less latency for the initial message to an actor and less GC pressure; whether there's a difference in later (especially after 10k+) messages to the same actor, idk.